src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs

```diff
@@ -189,32 +280,61 @@ public void Dispose()
 
         internal AbstractProject GetProject(ProjectId id)
         {
-            AbstractProject project;
-            _projectMap.TryGetValue(id, out project);
-            return project;
+            lock (_gate)
+            {
+                AbstractProject project;
+                _projectMap.TryGetValue(id, out project);
+                return project;
+            }
         }
 
         /// <summary>
         /// Add a project to the workspace.
+        /// If invoked on the foreground thread, the add is executed right away.
+        /// Otherwise, the add is scheduled on foreground task scheduler.
```

Comment:
UpdateProjectBinPath_Foreground is making calls back into AbstractProject instances for metadata to P2P reference conversion, and I didn't know what may go wrong if that happens from multiple background threads - I just retained that functionality on foreground thread.